### DISCUSSION TITLE
xarray and pandas accessors

### DISCUSSION THREAD

**Discussion by karosc** [microsoft/pylance-release#4564] [Q&A]
I am developing a scientific python packages that adds domain specific functionality to the xarray and pandas libraries. Per the [pandas docs](https://pandas.pydata.org/docs/development/extending.html#registering-custom-accessors) and  [xarray docs](https://docs.xarray.dev/en/stable/internals/extending-xarray.html) I have developed custom accessors that implement my desired functionality. 

Unfortunately, it doesn't seem like python language servers can read my accessors and enable autocompletion of methods and properties or display docstrings of the methods that I type out manually.

Has anyone thought of a solution to this problem that doesn't involve sub-classing the xarray or pandas data structures?

**Comment by rchiodo** [comment]
My guess is that this isn't possible without some changes to Pylance. You're essentially extending the DataFrame type at runtime. Pylance doesn't know anything about this happening without us adding special code to understand the `register_dataset_accessor` and `register_dataframe_accessor` decorators and what they modify. 

Perhaps if we supported merging type stubs (similar to the ask here: https://github.com/microsoft/pylance-release/issues/2866), you could add additional methods to the DataFrame type in a secondary type stub.

**Comment by karosc** [comment]
Thanks for the quick response @rchiodo!

I like the idea of a secondary stub, I'm not sure how I could get python (or pylance for that matter) to associate a secondary type stub with an xarray module if the stub itself isn't part of the xarray package. 

Do you think the pylance project be receptive to changes that would accommodate parsing these accessors that are set a runtime?

**Comment by rchiodo** [comment]
> Do you think the pylance project be receptive to changes that would accommodate parsing these accessors that are set a runtime?

Probably if we were open source, but unfortunately Pylance is not. I think the only thing possible externally is to up vote this discussion.

**Reply by rchiodo** [reply]
We did do something similar already for Pytest support. We treat `@pytest.fixture` specially so that Pylance understands pytest fixtures.

**Reply by karosc** [reply]
Do you think adding support for accessor auto completion and doc string parsing is indeed a pylance issue? Or it is something that can be done through the open source pyright package?

**Reply by rchiodo** [reply]
Generally yes. Pyright is really just a type checker and not a language service. Completions (especially for custom situations) are implemented in Pylance. 

Unless you can generate a consensus in the type checking community that this sort of thing should be supported by all type checkers, then Pyright would support it. 

For example, this PEP that was created for dataclass transform decorators:
https://peps.python.org/pep-0681/
